File: Sample Apps /SampleBlueDotExperience/app/src/main/java/com/mist/sample/bluedot/initializer/MistSdkManager.java
package com.mist.sample.bluedot.initializer;

import android.content.Context;
import android.util.Log;

import com.mist.android.ClientInformationCallback;
import com.mist.android.IndoorLocationCallback;
import com.mist.android.IndoorLocationManager;
import com.mist.android.VirtualBeaconCallback;

import java.lang.ref.WeakReference;

public class MistSdkManager{
    /**
     * Required by Mist SDK for initialization
     * IndoorLocationManager
     * IndoorLocationCallback
     * VirtualBeaconCallback
     */
    private IndoorLocationManager indoorLocationManager;
    private IndoorLocationCallback indoorLocationCallback;
    private VirtualBeaconCallback virtualBeaconCallback;

    private ClientInformationCallback clientInformationCallback;

    private static WeakReference<Context> contextWeakReference;
    private String envType, orgSecret;
    private static MistSdkManager sdkInitializer;

    private MistSdkManager() {
    }

    public static MistSdkManager getInstance(Context context) {
        contextWeakReference = new WeakReference<>(context);
        if (sdkInitializer == null) {
            sdkInitializer = new MistSdkManager();
        }
        return sdkInitializer;
    }

    public void init(String orgSecret, IndoorLocationCallback indoorLocationCallback, VirtualBeaconCallback virtualBeaconCallback, ClientInformationCallback clientInformationCallback) {
        if (orgSecret != null && !orgSecret.isEmpty()) {
            Log.d("", "SampleBlueDot init" + orgSecret);
            this.orgSecret = orgSecret;
            this.indoorLocationCallback = indoorLocationCallback;
            this.virtualBeaconCallback = virtualBeaconCallback;
            this.clientInformationCallback = clientInformationCallback;
        }
    }

    public synchronized void startMistSDK() {
        if (indoorLocationManager == null) {
            Log.d("", "indoorLocationManager start" + orgSecret);
            indoorLocationManager = IndoorLocationManager.getInstance(contextWeakReference.get(), orgSecret);
            indoorLocationManager.getClientInformation(clientInformationCallback);
            indoorLocationManager.setVirtualBeaconCallback(virtualBeaconCallback);
            indoorLocationManager.start(indoorLocationCallback);
        } else {
            restartMistSDK();
        }
    }

    public void stopMistSDK() {
        if (indoorLocationManager != null) {
            indoorLocationManager.stop();
        }
    }

    public void destroy() {
        if (indoorLocationManager != null) {
            indoorLocationManager.stop();
            indoorLocationManager = null;
        }
    }

    public void updateClientName(String clientName){
        indoorLocationManager.saveClientInformation(clientName,clientInformationCallback);
    }

    private synchronized void restartMistSDK() {
        if (indoorLocationManager != null) {
            stopMistSDK();
            indoorLocationManager.setVirtualBeaconCallback(virtualBeaconCallback);
            indoorLocationManager.getClientInformation(clientInformationCallback);
            indoorLocationManager.start(indoorLocationCallback);
        }
    }
}

File: Sample Apps /Kotlin/SampleBlueDotExperience/app/src/main/java/com/example/samplebluedotexperience/initializer/MistSdkManager.kt
package com.example.samplebluedotexperience.initializer

import android.content.Context
import android.util.Log
import com.mist.android.BatteryUsage
import com.mist.android.IndoorLocationCallback
import com.mist.android.IndoorLocationManager
import com.mist.android.external.config.LogLevel
import com.mist.android.external.config.MistConfiguration
import java.lang.ref.WeakReference

class MistSdkManager{

    /**
     * Required by Mist SDK for initialization
     * IndoorLocationManager
     * IndoorLocationCallback
     */

    private var indoorLocationManager: IndoorLocationManager?=null
    private lateinit var contextWeakReference: WeakReference<Context>
    private var mistConfiguration: MistConfiguration? = null
    private var sdkInitializer : MistSdkManager? = null
    private var orgSecret : String? = null
    private var indoorLocationCallback: IndoorLocationCallback? = null

    fun getInstance(context: Context): MistSdkManager {
        contextWeakReference = WeakReference<Context>(context)
        sdkInitializer?: run {
            sdkInitializer = MistSdkManager()
        }
        return sdkInitializer as MistSdkManager
    }

    fun init(orgSecret: String, indoorLocationCallback: IndoorLocationCallback, orgId: String) {
        this.indoorLocationCallback = indoorLocationCallback
        if (orgSecret.isNotEmpty()) {
            Log.d("", "Sample Blue Dot Init $orgSecret")
            this.orgSecret = orgSecret
            this.mistConfiguration =
                MistConfiguration(
                    context = contextWeakReference.get()!!,
                    token = orgSecret,
                    enableLog = true,
                    logLevel = LogLevel.INFO,
                    batteryUsage = BatteryUsage.HIGH_BATTERY_USAGE_HIGH_ACCURACY,
                    orgId = orgId
                )
        }

    }

        fun startMistSdk(){
            if(indoorLocationManager==null){
                indoorLocationManager = IndoorLocationManager
                mistConfiguration?.let {
                    indoorLocationCallback?.let { it1 ->
                        indoorLocationManager?.start(
                            it,
                            indoorLocationCallback = it1
                        )
                    }
                }
        }
        else{
            restartMistSdk()
        }
    }

    @Synchronized
    fun stopMistSdk(){
        indoorLocationManager?.let{
            indoorLocationManager?.stop()
        }
    }

    fun destroyMistSdk(){
        indoorLocationManager?.let{
            indoorLocationManager?.stop()
        }
    }

    @Synchronized
    private fun restartMistSdk() {
        indoorLocationManager?.let{
            stopMistSdk()
            mistConfiguration?.let {mistconfiguration ->
                indoorLocationCallback?.let { indoorlocationcallback ->
                    indoorLocationManager?.start(
                        mistconfiguration,
                        indoorLocationCallback = indoorlocationcallback
                    )
                }
            }
        }
    }

}